### ISSUE TITLE
Inconsistent symbols in RootOf

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#26808]
RootOfs can be made with different symbols but once a particular RootOf has been created with one symbol future ones will have the same symbol:
```python
In [1]: print(CRootOf(y**3 + y + 2, 1))
CRootOf(y**2 - y + 2, 0)

In [2]: print(CRootOf(x**3 + x + 1, 1))
CRootOf(x**3 + x + 1, 1)

In [3]: print(CRootOf(y**3 + y + 1, 1))
CRootOf(x**3 + x + 1, 1)

In [4]: print(CRootOf(x**3 + x + 2, 1))
CRootOf(y**2 - y + 2, 0)
```
This happens because of the cache and I think because `PurePoly` overrides `__eq__`.

This causes `xreplace` to fail:
```python
In [7]: print(CRootOf(y**3 + y + 2, 1))
CRootOf(y**2 - y + 2, 0)

In [8]: print(CRootOf(y**3 + y + 2, 1).xreplace({y:z}))
CRootOf(y**2 - y + 2, 0)
```
That in turn causes `as_dummy` to fail:
```python
In [10]: print(Integral(f(x), (x, 0, 1)).as_dummy())
Integral(f(_0), (_0, 0, 1))

In [11]: print(CRootOf(y**3 + y + 2, 1).as_dummy())
CRootOf(y**2 - y + 2, 0)
```
That causes `has_free` to fail:
```python
In [12]: print(CRootOf(y**3 + y + 2, 1).has_free(y))
True

In [13]: print(Integral(f(x), (x, 0, 1)).has_free(x))
False
```
That causes Poly to fail if the symbol matches the generator:
```python
In [14]: Poly(RootOf(x**3 + x + 1, 0), x)
...
PolynomialError: CRootOf(x**3 + x + 1, 0) contains an element of the set of generators.
```
In general I think that any time a Basic subclass overrides `__eq__` in a way that is not consistent with what `Basic.__eq__` would have done this is going to cause significant problems. I think that `PurePoly` should not be used in general or should be defined in such a way that it does not have any symbol:
https://github.com/sympy/sympy/blob/530149cc7256a98c5963bcccc43cec19a9d04d09/sympy/polys/polytools.py#L4425-L4447
PurePoly should just use numbered symbols like `#1`, `#2` like Mathematica's Root representation:
https://reference.wolfram.com/language/ref/Root.html

**Comment by asmeurer** [comment]
Yes, the "symbol" should exist just for the sake of printing, and for conversion to an expr. We should probably just call the symbols x1, x2, ..., and never actually store those. 

Is PurePoly actually intended to be an Expr? It only subclasses from Poly, and Poly subclasses from Basic, but honestly it really shouldn't. 

I'm honestly not even sure how I feel about PurePoly. It seems to mainly be used as a way to let methods that return polynomials like `charpoly` and `minpoly` avoid requiring a symbol from the user. I think you can make sense of it mathematically (it's basically an equivalence class), but it can't really be interpreted as a number, i.e., it shouldn't really be an Expr. But most places where you'd use `charpoly` or `minpoly` I think you do want an Expr. If you try to combine PurePoly with Expr you get nonsense like

```py
>>> PurePoly(x**2 + 1) + x
PurePoly(x**2 + x + 1, x, domain='ZZ')
```

Is there really a need for PurePoly to exist? SymPy already has a fairly robust way to create a unique symbol (`Dummy`). I think picking a unique representative of an equivalence class works better than trying to represent the whole class, because you don't have to deal with these equality issues. The downside is you always have to make sure your operations "factor through the class" (i.e., don't do something that is actually only dependent on the symbol you chose). But Dummy already takes care of most of the headaches that could arise from this. Plus, for user facing methods, you can just make the user specify a symbol, which they know will be unique. For equality you have to either normalize or use some sort of "dummy_eq" method. 

Is PurePoly used somewhere else I'm not aware of where you really need this sort of functionality?

**Comment by asmeurer** [comment]
Actually `minpoly` doesn't even use `PurePoly`

```py
>>> minpoly(sqrt(2))
_x**2 - 2
>>> type(minpoly(sqrt(2)))
<class 'sympy.core.add.Add'>
```

So it's only `charpoly` that does. I don't know if that was ever changed or if I'm just mixing up the two.

**Comment by oscarbenjamin** [comment]
> Actually `minpoly` doesn't even use `PurePoly`

```
In [8]: minpoly(sqrt(2), polys=True)
Out[8]: PurePoly(_x**2 - 2, _x, domain='QQ')

In [10]: type(minpoly(sqrt(2), polys=True).gens[0])
Out[10]: sympy.core.symbol.Dummy
```

It uses both PurePoly and a Dummy symbol.

**Comment by morluto** [comment]
Status check on current master (1.15.0.dev): the concrete inconsistencies reported here no longer reproduce. Instances are now canonicalized onto a shared dummy generator, which makes the reported symptoms disappear:

```python
>>> from sympy import *
>>> x, y, z = symbols('x y z')
>>> c = CRootOf(y**3 + y + 2, 1)
>>> c.poly
Poly(_x**2 - _x + 2, _x, domain='ZZ')     # shared _rootof_dummy
>>> c.free_symbols
set()
>>> c.has_free(y)
False                                     # was True when filed
>>> c.xreplace({y: z})
CRootOf(x**2 - x + 2, 0)                  # correct no-op: there is no y inside
```

Values are unaffected throughout: e.g. `CRootOf(y**3 + y + 1, 1)` and `CRootOf(x**3 + x + 1, 1)` return the same cached instance with identical numeric value, which is sound because PurePoly equality identifies polynomials up to generator renaming, and for univariate polynomials over the same exact domain a generator renaming is an isomorphism of the root sets.

The only remaining oddity is cosmetic: the string printer renders the shared dummy as a plain-looking `x`, so users see a symbol they never passed (as visible in the outputs above). If desired, printing dummies distinctively inside RootOf args (or documenting the canonicalization on the class) would avoid the confusion that led to this report.
